### DISCUSSION TITLE
Trying to understand WeConnectId login to make it work in c++

### DISCUSSION THREAD

**Discussion by JimKnopf1503** [robske110/IDDataLogger#31] [Q&A] [answered]
Hi!

First of all, thank you for you code and sharing your work!
I am writing a prgramm named SolarCharge, a all in one project.
It is working fine so far and I all already integrated my eGolf. Now we got an ID.4 and I am trying to make it also work but it dont.
I tryed to understand the code from iobroker adapter, but there are so many shortcuts like in php.
I have troubel after sending the password. I got finaly a 401 response with the message in the body: error":"invalid_token".
The redirect befor contain all the tokens needed but the tokenexchange fail with code 401: Error: Please provide a valid auth token.
My code for parsing:
if((Identities.relayState=parse(Form,"input","name","relayState"))=="") return -104; if((Identities.relayState=parse(Identities.relayState,"value"))=="") return -105; if((CSRF=parse(Form,"input","name","_csrf"))=="") return -106; if((CSRF=parse(CSRF,"value"))=="") return -107; if((Identities.hmac=parse(Form,"input","name","hmac"))=="") return -108; if((Identities.hmac=parse(Identities.hmac,"value"))=="") return -109;
The code for the request:
optionString="Content-Type: application/x-www-form-urlencoded§$user-agent: SolarCharge§$Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3§$Accept-Language: en-US,en;q=0.9§$Accept-Encoding: gzip, deflate§$x-requested-with: ";// §$ =separator for request optionString+=xrequest; sPostfields = ("email=" + WeConnectValues->User+"&password="+WeConnectValues->PassWd +"&relayState="+Identities.relayState+ "&hmac="+Identities.hmac+"&_csrf="+CSRF); request.followRedirect=true; url=LOGIN_HANDLER_BASE+action; request.request(optionString,url,sPostfields);

Did I name something wrong? I cant verify because in your code the values are not shown.
In hope for help,
Burkhard

**Comment by robske110** [comment, marked as answer]
Hi! Unfortunately I can not find anything wrong in the 2 code snippets you sent, at least not without context. I am also not very proficient in C++.
I'd strongly recommend using a library like curl (which IDDataLogger is also doing), which massively simplifies things.
I obviously don't know the specifics of your project (for example if it is supposed to be something embedded or public), but if it is just a home project another option would be to run IDDataLogger and query its API, which is very straightforward to do. (It's just a simple HTTP request with a static api key in the URL as a query parameter and it returns JSON)
See: https://github.com/robske110/IDDataLogger/wiki/API-reference#carstatusphp

**Comment by JimKnopf1503** [comment]
Hi!

Thanks for your answer.
My project will be open source when I reached a better state so I can publish it.
It is a litte more complex, combining my energystorrige, photovoltaik and two different cars (eGolf and ID.4) and a Keba P30 and a PhonixCantact based wallbox. It will in a first step take up to 8 cars and up to 8 wallboxes.
I would like to keep it completly in c++ to keep it fast and more easy to install.
I am using CURL but not something ling REQUEST. I named my object request, which is setting the curl-options and perfomr the requests, like your wrapper. I started this project 4 years ago and it worked until I got the second car.

Edit: I made it. I had trouble to understand parts of MobileAppApi.php.
Line 16 : const API_BASE = "https://mobileapi.apps.emea.vwapps.io";
Line 110: public function apiPost(string $apiEndpoint, $body, string $apiBase = self::API_BASE, ?array $header = null) .......
But in reality "https://login.apps.emea.vwapps.io" is used as API_BASE at "Getting real token..."  and I can't see where it is changed.
But I don't need to know. I think I can login right now. Now going at the data.

Thanks for your code and help!
Burkhard
